src/lib/spinel/spinel_driver.cpp

```diff
@@ -0,0 +1,466 @@
+ *  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ *  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ *  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ *  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ *  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "spinel_driver.hpp"
+
+#include <assert.h>
+
+#include <openthread/platform/time.h>
+
+#include "common/code_utils.hpp"
+#include "common/new.hpp"
+#include "common/num_utils.hpp"
+#include "lib/platform/exit_code.h"
+#include "lib/spinel/spinel.h"
+
+namespace ot {
+namespace Spinel {
+
+constexpr spinel_tid_t sTid = 1; ///< In Spinel Driver, only use Tid as 1.
+
+inline bool SpinelDriver::IsFrameForUs(spinel_iid_t aIid)
+{
+    bool found = false;
+
+    for (spinel_iid_t iid : mIidList)
+    {
+        if (aIid == iid)
+        {
+            ExitNow(found = true);
+        }
+    }
+
+exit:
+    return found;
+}
+
+SpinelDriver::SpinelDriver(void)
+    : Logger("SpinelDriver")
+    , mSpinelInterface(nullptr)
+    , mWaitingKey(SPINEL_PROP_LAST_STATUS)
+    , mIsWaiting(false)
+    , mIid(SPINEL_HEADER_INVALID_IID)
+    , mSpinelVersionMajor(-1)
+    , mSpinelVersionMinor(-1)
+    , mIsReady(false)
+{
+    memset(mIidList, SPINEL_HEADER_INVALID_IID, sizeof(mIidList));
+    memset(mVersion, 0, sizeof(mVersion));
+
+    mReceivedFrameHandler.Set(&HandleInitialFrame, this);
+}
+
+void SpinelDriver::Init(SpinelInterface    &aSpinelInterface,
+                        bool                aSwReset,
+                        const spinel_iid_t *aIidList,
+                        uint8_t             aIidListLength)
+{
+    otError error = OT_ERROR_NONE;
+
+    mSpinelInterface = &aSpinelInterface;
+    mRxFrameBuffer.Clear();
+    SuccessOrDie(mSpinelInterface->Init(HandleReceivedFrame, this, mRxFrameBuffer));
+
+    VerifyOrDie(aIidList != nullptr, OT_EXIT_INVALID_ARGUMENTS);
+    VerifyOrDie(aIidListLength != 0 && aIidListLength <= OT_ARRAY_LENGTH(mIidList), OT_EXIT_INVALID_ARGUMENTS);
+    mIid = aIidList[0];
+    memset(mIidList, SPINEL_HEADER_INVALID_IID, sizeof(mIidList));
+    memcpy(mIidList, aIidList, aIidListLength * sizeof(spinel_iid_t));
+
+    ResetCp(aSwReset);
+    SuccessOrExit(error = CheckSpinelVersion());
+    SuccessOrExit(error = GetCpVersion());
+
+    // TODO: Check Capability and decide the mode
+
+exit:
+    SuccessOrDie(error);
+}
+
+void SpinelDriver::Deinit(void)
+{
+    // This allows implementing pseudo reset.
+    new (this) SpinelDriver();
+}
+
+void SpinelDriver::Recover(bool aSkipReset)
```

Comment:
Could we split this function into two functions `ClearRxBuffer()` and `SetCoProcessorReady()`? I think this could make it more flexible and improves readability.